### PR TITLE
[main] Introduce cmdline option parser to share some code between executables

### PR DESCRIPTION
**Pull request by silverweed**
We now have enough C++ executables (with more to be ported) to justify sharing some code between them, in particular option parsing.

This PR introduces a simple option parser class that covers most of our cases. Not all cases (e.g. `hadd` keeps its custom parsing because it's a bit weird), but it's already enough to cover `rootls` and `rootbrowse`, and will in the future also cover `rootcp`, `rootmv` etc.

The parser is documented with code examples and properly tested and it's about 300 lines of codes (about an order of magnitude less than e.g. [cxxopts](https://github.com/jarro2783/cxxopts/) which in my opinion is overkill for our use case).

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

File: main/src/optparse.hxx [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+/// MyAppOpts ParseArgs(const char **args, int nArgs) {
+///    ROOT::RCmdLineOpts opts;
+///    // will parse '-c VAL', '--compress VAL' or '--compress=VAL'
+///    opts.AddFlag({"-c", "--compress"}, RCmdLineOpts::EFlagType::kWithArg);
+///    // will toggle a boolean flag '--recreate' (no args).
+///    opts.AddFlag({"--recreate"});
+///    opts.AddFlag({"-o"}, RCmdLineOpts::EFlagType::kWithArg);
+///
+///    // NOTE: `args` should not contain the program name! It should usually be `argc + 1`.
+///    opts.Parse(args, nArgs);
```
Comment (pcanal) [inline main/src/optparse.hxx:18]:
```suggestion
///    // NOTE: `args` should not contain the program name! It should usually be `argc + 1`.
///    // For example `main(char **argv, int argc)` might call this function as: 
///    //    `ParseArgs(const_cast<const_cast<const char**>(argv) + 1, argc - 1);`
///    opts.Parse(args, nArgs);
```
(Without this, I was side tracked for a while pondering whether the comment wanted the last line to be `opts.Parse(args, nArgs+1)` )

File: main/src/rootls.cxx [hunk 2]
```diff
@@ -784,9 +746,9 @@ int main(int argc, char **argv)
    gErrorIgnoreLevel = kError;
 
    auto args = ParseArgs(const_cast<const char **>(argv) + 1, argc - 1);
```
Comment (pcanal) [inline main/src/rootls.cxx:748]:
```suggestion
   auto args = ParseArgs(argv + 1, argc - 1);
```
I.e. is the `const_cast` still needed?

Comment (pcanal) [reply main/src/rootls.cxx:748]: Nevermind, I mis-read the direction of the `const_cast` (it is adding the `const`)

File: main/test/optparse_test.cxx [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+   EXPECT_TRUE(opts.GetBooleanFlag("f"));
+   EXPECT_FALSE(opts.GetBooleanFlag("g"));
+}
+
+TEST(OptParse, OptParseString)
+{
+   ROOT::RCmdLineOpts opts;
+   opts.AddFlag({"--foo"}, ROOT::RCmdLineOpts::EFlagType::kWithArg);
+
+   const char *args[] = {"--foo", "bar"};
```
Comment (pcanal) [inline main/test/optparse_test.cxx:58]: Do we test something like:
```
 const char *args[] = {"somename", "--foo", "bar", "-abc"};
```
and
```
 const char *args[] = {"somename", "--noargopt", "bar"};
```
```
 const char *args[] = {"--noargopt", "bar"};
```
i.e. location of positional arguments.  (In particular the first one might appear by mistake if `argv` is passed instead of `argv+1`)

**Comment by ferdymercury** [comment]
Thanks a lot for the initiative!

Even if it's 300 lines of code, isn't it better to rely on something external that gets auto-maintained, even if it's bigger in number of lines of code?
Such as the header-only https://github.com/CLIUtils/CLI11 by @henryiii

Or to understand better the motivation, what do we gain by having a small parser for which we have to add our own tests?
- Is it because then compile time is smaller?
- Or is initialization time more performant?
- Or is the overkill of CLI11 or cxxopt increasing the size of the produced binaries or sth like that?
- Or to not rely on external sources?

**Comment by silverweed** [comment]
> Thanks a lot for the initiative!
> 
> Even if it's 300 lines of code, isn't it better to rely on something external that gets auto-maintained, even if it's bigger in number of lines of code? Such as the header-only https://github.com/CLIUtils/CLI11
> 
> Or to understand better the motivation, what do we gain by having a small parser for which we have to add our own tests?
> 
>     * Is it because then compile time is smaller?
> 
>     * Or is initialization time more performant?
> 
>     * Or is the overkill of CLI11 or cxxopt increasing the size of the produced binaries or sth like that?
> 
>     * Or to not rely on external sources?

It's mostly the last point (though the first is also somewhat important).
Simply put, if the functionality is not particularly complex (as is this case) it's always better to have your own code that you know what it does and can easily control than an external dependency that you need to maintain and complicates your build system. 
Furthermore, if such dependency is 10x the code size (without providing at least 10x the benefits) then you would just be carrying around much more weight than you really need. 
External dependencies are not free and should only be used where there is a really good argument for them, which in this case I don't see.

File: main/test/optparse_test.cxx (hunk 3 above)
Comment (silverweed) [reply main/test/optparse_test.cxx:58]: currently positional arguments may appear in any place, so in the first example `somename` would be the first (and only) positional arg, in the second `somename` and `bar` would be the positional args and in the third only `bar`.
I will add the tests for the first 2 cases (as the third I believe is already covered)

**Comment by henryiii** [comment]
On the flipside, a library that is heavily used is more likely to come across all of the weird edge cases that are going to end up making the custom implementation significantly longer eventually.

When I designed CLI11, I did have ROOT in mind, and with that you could actually make it easy to have a user extend an application. That would definitely be more than 10x functionality.

**Comment by silverweed** [comment]
> weird edge cases that are going to end up making the custom implementation significantly longer eventually.

This is an unprovable assumption as of now - if we start finding such edge cases regularly then this might tilt the decision towards an external library. It's very easy to change our mind later after all, this is all "leaf code", which brings me to the second point:

>you could actually make it easy to have a user extend an application

I don't see the point here: the user cannot "extend" a binary like `rootls` or `rootcp` as in "use it as a library", and if they want to write some sort of wrapper around them they are free to bring their own option parsing library. `optparse.hxx` is not meant to be used as a library, which is why it's in `src` and not in `inc/ROOT`.

> That would definitely be more than 10x functionality.

Again, this is conjecture which can only be proven with actually trying both things. My position is: let's start with the simpler option, then if we realize it was a mistake we can easily swap the implementation later.

File: main/src/optparse.hxx [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+///
+/// Small utility to parse cmdline options.
+///
+/// Usage:
+/// ~~~{.cpp}
+/// MyAppOpts ParseArgs(const char **args, int nArgs) {
+///    ROOT::RCmdLineOpts opts;
+///    // will parse '-c VAL', '--compress VAL' or '--compress=VAL'
+///    opts.AddFlag({"-c", "--compress"}, RCmdLineOpts::EFlagType::kWithArg);
+///    // will toggle a boolean flag '--recreate' (no args).
```
Comment (jblomer) [inline main/src/optparse.hxx]:
```suggestion
///    // will toggle a switch '--recreate' (no args).
```

File: main/src/optparse.hxx [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace ROOT {
+
+class RCmdLineOpts {
+public:
+   enum class EFlagType {
+      kBoolean,
```
Comment (jblomer) [inline main/src/optparse.hxx]:
```suggestion
      kSwitch,
```

File: main/src/optparse.hxx [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+      kWithArg
+   };
+
+   struct RFlag {
+      std::string fName;
+      std::string fValue;
```
Comment (jblomer) [inline main/src/optparse.hxx:78]: Should this be an optional to distinguish a switch from an empty string argument?

File: main/src/optparse.hxx [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+      for (std::size_t i = 0; i < nArgs && fErrors.empty(); ++i) {
+         const char *arg = args[i];
+
+         if (strcmp(arg, "--") == 0) {
+            forcePositional = true;
+            continue;
+         }
+
+         bool isFlag = !forcePositional && arg[0] == '-';
+         if (isFlag) {
```
Comment (jblomer) [inline main/src/optparse.hxx]: Maybe swap the if-else blocks so that the short one is handled first (reduce scrolling).

File: main/src/optparse.hxx [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+      for (const auto &err : fErrors)
+         stream << err << "\n";
+      return !fErrors.empty();
+   }
+
+   void AddFlag(std::initializer_list<std::string_view> aliases, EFlagType type = EFlagType::kBoolean,
+                std::string_view help = "")
+   {
+      int aliasIdx = -1;
+      for (auto f : aliases) {
```
Comment (jblomer) [inline main/src/optparse.hxx:136]: Perhaps: check that the alias is neither `-` nor contains a space (and possibly other funny stuff, newline, NULL, etc). Maybe enough to say in the documentation what is allowed.

**Review by jblomer — APPROVED** [review]
Nice! Two general comments:
  - I'd call "boolean flag" a switch
  - Perhaps call out in the initial documentation that repeated flags are not supported (e.g., no `root-ssh -vvv`)

File: main/src/optparse.hxx (hunk 6 above)
Comment (silverweed) [reply main/src/optparse.hxx:78]: Is the distinction relevant? As far as I see it, it would only be relevant if the user is iterating the flags list and wants to know which flag is a switch and which is not; but the user *knows* the semantic value of the flags already (as he defined them), so my impression is that it's fine to collapse the two cases.
But maybe I'm missing some other benefits to this, feel free to let me know if you have any in mind

File: main/src/optparse.hxx (hunk 8 above)
Comment (silverweed) [reply main/src/optparse.hxx:136]: The alias already cannot be `-` (or start with a `-` after the prefix) as that is prevented by the combination of the two `if`s; I could validate the characters, but I think it's a bit overly defensive (the user will almost always define these aliases as literal strings and probably won't shoot themselves in the foot by using weird characters in flags)
